windows/src/engine/keyman32/preservedkeymap.cpp

```diff
@@ -318,13 +370,114 @@ BOOL PreservedKeyMap::MapKeyboard(KEYBOARD *pKeyboard, PreservedKey **pPreserved
+
+  // This is where we will call down to the api to get list of keys used in the keyboard rules
+  km_kbp_keyboard_key *kb_key_list;
+
+  km_kbp_status err_status = km_kbp_keyboard_get_key_list(pKeyboard, &kb_key_list);
+  if ((err_status != KM_KBP_STATUS_OK) || (kb_key_list ==nullptr)) {
+    return FALSE;
+  }
+
+  km_kbp_keyboard_key *key_rule_it = kb_key_list;
+  for (; key_rule_it->key; ++key_rule_it) {
+    ++cRules;
+  }
+  cKeys = cRules;
+  if (cKeys == 0) {
+    km_kbp_keyboard_key_list_dispose(kb_key_list);
+    return FALSE;
+  }
+
+  if (bSimulateAltGr) {
+    // We might need twice as many preserved keys to map both LCtrl+LAlt+x and RAlt+x
+    cKeys *= 2;
+  }
+
+  if (pPreservedKeys == NULL) {
+    *cPreservedKeys = cKeys;
+    km_kbp_keyboard_key_list_dispose(kb_key_list);
+    return TRUE;
+  }
+
+  if (*cPreservedKeys < cKeys) {
+    km_kbp_keyboard_key_list_dispose(kb_key_list);
+    return FALSE;
+  }
+
+  PreservedKey *pKeys = *pPreservedKeys;
+
+
+  for (i = 0; i < cRules; i++) {
+    // If we have a key rule for the key, we should preserve it
+    if (MapKeyRuleCore(&kb_key_list[i], &pKeys[n].key)) {
+      // Don't attempt to add the same preserved key twice. Bad things happen
+      if (!IsMatchingKey(&pKeys[n], pKeys, n)) {
+        CoCreateGuid(&pKeys[n].guid);
+        n++;
+
+        if (bSimulateAltGr && (pKeys[n - 1].key.uModifiers & RALT_MATCHING_MASK) == TF_MOD_RALT) {
+          // Do this for RALT and RALT+SHIFT only, so we've tested against that mask
+          // Copy the key and fix modifiers
+          pKeys[n].key            = pKeys[n - 1].key;
+          pKeys[n].key.uModifiers = (pKeys[n].key.uModifiers & ~TF_MOD_RALT) | TF_MOD_LCONTROL | TF_MOD_LALT;
+          CoCreateGuid(&pKeys[n].guid);
+          n++;
+        }
+      }
+    }
+  }
+
+  km_kbp_keyboard_key_list_dispose(kb_key_list);
+
+  *cPreservedKeys = n;  // return actual count of allocated keys, usually smaller than allocated count
+  return TRUE;
+}
+
+
+
+
+
 extern "C" __declspec(dllexport) BOOL WINAPI GetKeyboardPreservedKeys(PreservedKey **pPreservedKeys, size_t *cPreservedKeys)
 {
   PreservedKeyMap pkm;
   PKEYMAN64THREADDATA _td = ThreadGlobals();
-  if(!_td) return FALSE;
-  if(!_td->lpActiveKeyboard) return FALSE;
-  if(!_td->lpActiveKeyboard->Keyboard) return FALSE;
+  if (!_td) {
+    return FALSE;
+  }
+  if (!_td->lpActiveKeyboard) {
+    return FALSE;
+  }
+  // It could be an active core keyboard
+  if (Globals::get_CoreIntegration()) {
+
+    if (!_td->lpActiveKeyboard->lpCoreKeyboard) {
+      return FALSE;
+    }
+    // use api to get key rules
+    return pkm.MapKeyboardCore(_td->lpActiveKeyboard->lpCoreKeyboard, pPreservedKeys, cPreservedKeys);
+
+  } else {
+    if (!_td->lpActiveKeyboard->Keyboard) {
+      return FALSE;
+    }
+    return pkm.MapKeyboard(_td->lpActiveKeyboard->Keyboard, pPreservedKeys, cPreservedKeys);
+  }
```

Comment:
Do we want to add the issue number for future removal of this code and related functions?